### PR TITLE
[dhcp-relay] make DHCP relay an extension

### PR DESCRIPTION
**Pull request by stepanblyschak**
**- Why I did it**
Make DHCP relay docker an extension. DHCP relay now carries dhcp relay commands CLI plugin and has a complete manifest.
It is installed as extension if INCLUDE_DHCP_REALY is set to y.

DEPENDS on https://github.com/Azure/sonic-buildimage/pull/5939

**- How I did it**
Modify DHCP relay docker makefile and dockerfile. Make changes to sonic_debian_extension.j2 to install sonic packages.
I moved DHCP related CLI tests from sonic-utilities to DHCP relay docker.
This PR introduces a way to write a plugin as part of docker image and run the tests from cli-plugin-tests directory under docker directory.
The test result is available in target/docker-dhcp-relay.gz.log:
```
[ REASON ] :      target/docker-dhcp-relay.gz does not exist   NON-EXISTENT PREREQUISITES: docker-start target/docker-config-engine-buster.gz-load target/python-wheels/sonic_utilities-1.2-py3-none-any.whl-in
stall target/debs/buster/python3-swsscommon_1.0.0_amd64.deb-install
[ FLAGS  FILE    ] : []
[ FLAGS  DEPENDS ] : []
[ FLAGS  DIFF    ] : []
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /sonic/dockers/docker-dhcp-relay/cli-plugin-tests, inifile:
plugins: cov-2.6.0
collecting ... collected 10 items

test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_plugin_registration PASSED [ 10%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_add_dhcp_relay_with_nonexist_vlanid PASSED [ 20%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_add_dhcp_relay_with_invalid_vlanid PASSED [ 30%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_add_dhcp_relay_with_invalid_ip PASSED [ 40%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_add_dhcp_relay_with_exist_ip PASSED [ 50%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_add_del_dhcp_relay_dest PASSED [ 60%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_remove_nonexist_dhcp_relay_dest PASSED [ 70%]
test_config_dhcp_relay.py::TestConfigVlanDhcpRelay::test_config_vlan_remove_dhcp_relay_dest_with_nonexist_vlanid PASSED [ 80%]
test_show_dhcp_relay.py::TestVlanDhcpRelay::test_plugin_registration PASSED [ 90%]
test_show_dhcp_relay.py::TestVlanDhcpRelay::test_dhcp_relay_column_output PASSED [100%]

=============================== warnings summary ===============================
/usr/local/lib/python3.7/dist-packages/tabulate.py:7
  /usr/local/lib/python3.7/dist-packages/tabulate.py:7: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import namedtuple, Iterable

-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================== 10 passed, 1 warnings in 0.35 seconds =====================

```

**- How to verify it**
Build it with and without INCLUDE_DHCP_RELAY.

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by stepanblyschak** [comment]
@renukamanavalan @jleveque @tahmed-dev @lguohan @bandaru-viswanath

File: dockers/docker-dhcp-relay/cli/show/plugins/dhcp_relay.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    vlan_config = vlan_dhcp_helper_data.get(vlan)
+    if not vlan_config:
+        return ""
+
+    dhcp_helpers = vlan_config.get('dhcp_servers', [])
+
+    return '\n'.join(natsorted(dhcp_helpers))
+
+
+vlan.VlanBrief.COLUMNS.append(('DHCP Helper Address', get_dhcp_helper_address))
```
Comment (bandaru-viswanath) [inline dockers/docker-dhcp-relay/cli/show/plugins/dhcp_relay.py:17]: Would this be problematic, plugins changing parent data structures ?  There probably needs to be better mechanisms for this ?

File: dockers/docker-nat/Dockerfile.j2 [hunk 2]
```diff
@@ -2,6 +2,8 @@
 FROM docker-config-engine-buster
 
 ARG docker_container_name
+ARG manifest
```
Comment (bandaru-viswanath) [inline dockers/docker-nat/Dockerfile.j2:5]: Should the J2 file changes be in a different PR, let this PR be just the dhcp-relay changes.

File: rules/docker-dhcp-relay.mk [hunk 3]
```diff
@@ -15,14 +15,38 @@ $(DOCKER_DHCP_RELAY)_DBG_IMAGE_PACKAGES = $($(DOCKER_CONFIG_ENGINE_BUSTER)_DBG_I
...
 SONIC_DOCKER_DBG_IMAGES += $(DOCKER_DHCP_RELAY_DBG)
-SONIC_INSTALL_DOCKER_DBG_IMAGES += $(DOCKER_DHCP_RELAY_DBG)
+
+ifeq ($(INCLUDE_KUBERNETES),y)
+$(DOCKER_DHCP_RELAY)_DEFAULT_FEATURE_OWNER = kube
+endif
+
+$(DOCKER_DHCP_RELAY)_DEFAULT_FEATURE_STATE_ENABLED = y
+
+ifeq ($(INSTALL_DEBUG_TOOLS),y)
```
Comment (bandaru-viswanath) [inline rules/docker-dhcp-relay.mk:36]: Just a clarification for me, not a review comment - Should 'INSTALL_DEBUG_TOOLS' necessarily mean the debug images/dockers ?

**Review by bandaru-viswanath — CHANGES_REQUESTED** [review]
Overall, the feedback is that -
a) This PR has changes beyond the dhcp-relay becoming an 'extension', all the j2 and mk file changes apart from the dhcp-relay can be moved into an independent PR

b) Specifically for the CLI, I am apprehensive of some of the implicit knowledge of the plugin and parent (vlan in this case). It would be good to make these as explicit.

File: rules/docker-dhcp-relay.mk (hunk 3 above)
Comment (stepanblyschak) [reply rules/docker-dhcp-relay.mk:36]: Yes, as you can see for regular dockers it means we will install -dbg versions if INSTALL_DEBUG_TOOLS=y
https://github.com/Azure/sonic-buildimage/blob/master/platform/vs/one-image.mk#L8

File: dockers/docker-nat/Dockerfile.j2 (hunk 2 above)
Comment (stepanblyschak) [reply dockers/docker-nat/Dockerfile.j2:5]: These changes are in https://github.com/Azure/sonic-buildimage/pull/5939.
It is not simple to generate a PR which touches changes done in other PR.
So after https://github.com/Azure/sonic-buildimage/pull/5939 is merged this diff will be shown as only DHCL relay changes.

File: dockers/docker-dhcp-relay/cli/show/plugins/dhcp_relay.py (hunk 1 above)
Comment (stepanblyschak) [reply dockers/docker-dhcp-relay/cli/show/plugins/dhcp_relay.py:17]: Would it be better if VlanBrief had a public class method:
```python
def register_column(column_name, column_data_callback)
```
And a comment that this API is in use by plugins to register additional columns?

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+from click.testing import CliRunner
+
+from utilities_common.db import Db
+
+import pytest
+
+sys.path.append('../cli/config/plugins/')
+import dhcp_relay
+
+@pytest.fixture()
```
Comment (tahmed-dev) [inline dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:15]: this fixture is already defined in the conftest.py.

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    yield cfgdb
+
+
+config_vlan_add_dhcp_relay_output="""\
+Added DHCP relay destination address 192.0.0.100 to Vlan1000
+Restarting DHCP relay service...
+"""
+
+config_vlan_del_dhcp_relay_output="""\
```
Comment (tahmed-dev) [inline dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:44]: I think `\` is not required here.

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+Removed DHCP relay destination address 192.0.0.100 from Vlan1000
+Restarting DHCP relay service...
+"""
+
+class TestConfigVlanDhcpRelay(object):
+    @classmethod
+    def setup_class(cls):
+        os.environ['UTILITIES_UNIT_TESTING'] = "1"
```
Comment (tahmed-dev) [inline dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:52]: What is the use of `UTILITIES_UNIT_TESTING` environment variable?

File: rules/docker-dhcp-relay.mk [hunk 7]
```diff
@@ -17,15 +17,36 @@ $(DOCKER_DHCP_RELAY)_LOAD_DOCKERS = $(DOCKER_CONFIG_ENGINE_BUSTER)
...
-$(DOCKER_DHCP_RELAY)_RUN_OPT += --privileged -t
-$(DOCKER_DHCP_RELAY)_RUN_OPT += -v /etc/sonic:/etc/sonic:ro
-$(DOCKER_DHCP_RELAY)_RUN_OPT += -v /usr/share/sonic/scripts:/usr/share/sonic/scripts:ro
+$(DOCKER_DHCP_RELAY)_CONTAINER_PRIVILEGED = true
+$(DOCKER_DHCP_RELAY)_CONTAINER_VOLUMES += /etc/sonic:/etc/sonic:ro
+$(DOCKER_DHCP_RELAY)_CONTAINER_VOLUMES += /usr/share/sonic/scripts:/usr/share/sonic/scripts:ro
+$(DOCKER_DHCP_RELAY)_CONTAINER_TMPFS += /tmp/
+$(DOCKER_DHCP_RELAY)_CONTAINER_TMPFS += /var/tmp/
+
+$(DOCKER_DHCP_RELAY)_CLI_CONFIG_PLUGIN = /cli/config/plugins/dhcp_relay.py
```
Comment (tahmed-dev) [inline rules/docker-dhcp-relay.mk:49]: how are the cli scripts being exposed to the host os?

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py (hunk 5 above)
Comment (stepanblyschak) [reply dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:44]: \ here is to not have newline as first character in this string otherwise this expected output will not match exactly.

File: rules/docker-dhcp-relay.mk (hunk 7 above)
Comment (stepanblyschak) [reply rules/docker-dhcp-relay.mk:49]: When sonic-package-manager installs the docker and sees that there are CLI plugins it will install CLI plugins from docker to the host.

Comment (tahmed-dev) [reply rules/docker-dhcp-relay.mk:49]: I see. The only concern I had here is that were exposing files from container to host which in a sense goes against containerization.

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            print(result.output)
+            assert result.exit_code == 0
+            assert result.output == config_vlan_add_dhcp_relay_output
+            assert mock_run_command.call_count == 3
+            db.cfgdb.set_entry.assert_called_once_with('VLAN', 'Vlan1000', {'dhcp_servers': ['192.0.0.1', '192.0.0.100']})
+
+        db.cfgdb.set_entry.reset_mock()
+
+        # del relay dest
+        with mock.patch("utilities_common.cli.run_command") as mock_run_command:
```
Comment (tahmed-dev) [inline dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:103]: most of the test cases can parameterized. Can we convert them into parameterized test case.

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py (hunk 4 above)
Comment (stepanblyschak) [reply dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py]: Removed

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py (hunk 6 above)
Comment (stepanblyschak) [reply dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py]: Removed

File: rules/docker-dhcp-relay.mk (hunk 7 above)
Comment (stepanblyschak) [reply rules/docker-dhcp-relay.mk:49]: Yes, you have a valid concern. This was a topic of discussions in SONiC App Ext & Kubernetes subgroup. For phase 1 we defined this simple approach. We aim for CLI autogeneration to solve this.

File: dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py (hunk 8 above)
Comment (stepanblyschak) [reply dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:103]: This moved from sonic-utilities tests mostly as is. IMO this is a bit different effort. Let me know if you think it is required here.

Comment (tahmed-dev) [reply dockers/docker-dhcp-relay/cli-plugin-tests/test_config_dhcp_relay.py:103]: I see. did not realize they were moved from sonic-utilities.

**Comment by tahmed-dev** [comment]
@stepanblyschak , thanks for your efforts. Wonder if you would please update the PR description of these tests are run and how to view the coverage report. Would package build fail if any test fail?

File: files/build_templates/init_cfg.json.j2 [hunk 9]
```diff
@@ -46,7 +45,7 @@
             "has_per_asic_scope": {% if feature + '@.service' in installer_services.split(' ') %}true{% else %}false{% endif %},
             "auto_restart": "{{autorestart}}",
 {%- if include_kubernetes == "y" %}
-{%- if feature in ["dhcp_relay", "lldp", "pmon", "radv", "snmp", "telemetry"] %}
+{%- if feature in ["lldp", "pmon", "radv", "snmp", "telemetry"] %}
```
Comment (tahmed-dev) [inline files/build_templates/init_cfg.json.j2:48]: @renukamanavalan please review this change regarding dhcp_relay and kubernetes.

File: files/scripts/swss.sh [hunk 10]
```diff
@@ -1,8 +1,17 @@
 #!/bin/bash
 
-DEPENDENT="radv dhcp_relay"
+DEPENDENT="radv"
 MULTI_INST_DEPENDENT="teamd"
 
+# Update dependent list based on other packages requirements
```
Comment (tahmed-dev) [inline files/scripts/swss.sh:6]: is this change related to this dhcp_relay PR? It seems to me it should apply to every feature that is moving to appxt.

File: rules/sonic-packages.mk [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# rules to define remote packages that need to be installed
```
Comment (tahmed-dev) [inline rules/sonic-packages.mk:1]: is this related?

File: files/build_templates/sonic_debian_extension.j2 [hunk 12]
```diff
@@ -654,6 +654,28 @@ sudo https_proxy=$https_proxy LANG=C chroot $FILESYSTEM_ROOT docker pull k8s.gcr
 echo "docker images pull complete"
 {% endif %}
 
+{% macro get_install_options(set_owner, enabled) -%}
```
Comment (tahmed-dev) [inline files/build_templates/sonic_debian_extension.j2:657]: This section seems to be generic. I suggest slitting this PR into two:
1. appxt Infra work
2. porting dhcp_relay to appxt.

File: slave.mk [hunk 13]
```diff
@@ -747,6 +751,7 @@ $(addprefix $(TARGET_PATH)/, $(DOCKER_IMAGES)) : $(TARGET_PATH)/%.gz : .platform
 		$(eval export $(subst -,_,$(notdir $($*.gz_PATH)))_whls=$(shell printf "$(subst $(SPACE),\n,$(call expand,$($*.gz_PYTHON_WHEELS)))\n" | awk '!a[$$0]++'))
 		$(eval export $(subst -,_,$(notdir $($*.gz_PATH)))_dbgs=$(shell printf "$(subst $(SPACE),\n,$(call expand,$($*.gz_DBG_PACKAGES)))\n" | awk '!a[$$0]++'))
 		$(eval export $(subst -,_,$(notdir $($*.gz_PATH)))_pkgs=$(shell printf "$(subst $(SPACE),\n,$(call expand,$($*.gz_APT_PACKAGES)))\n" | awk '!a[$$0]++'))
+		if [ -d $($*.gz_PATH)/cli-plugin-tests/ ]; then pushd $($*.gz_PATH)/cli-plugin-tests; pytest-$($(SONIC_UTILITIES_PY3)_PYTHON_VERSION) -v $(LOG); popd; fi
```
Comment (tahmed-dev) [inline slave.mk:754]: this should be part of infra work.

**Comment by stepanblyschak** [comment]
@tahmed-dev I see, so you suggest to split this PR,. Ok, that makes sense. Thanks you

**Comment by stepanblyschak** [comment]
@tahmed-dev PR: https://github.com/Azure/sonic-buildimage/pull/7593. This PR includes it, but once it is merged this PR will have a diff only relevant to dhcp-relay.

**Review by tahmed-dev — DISMISSED** [review]
LGTM. Please merge after #7593 is merged.

**Comment by stepanblyschak** [comment]
```
tests/hostcfgd/hostcfgd_test.py:5: in <module>
    from parameterized import parameterized
E   ModuleNotFoundError: No module named 'parameterized'

```
Same build error observed in different PR - https://dev.azure.com/mssonic/build/_build/results?buildId=19675&view=logs&j=88ce9a53-729c-5fa9-7b6e-3d98f2488e3f&t=8d99be27-49d0-54d0-99b1-cfc0d47f0318

**Comment by qiluo-msft** [comment]
Please check if https://github.com/Azure/sonic-buildimage/pull/7900 fixed the build.

**Comment by stepanblyschak** [comment]
Still issues with build same as https://github.com/Azure/sonic-buildimage/pull/7848#issuecomment-859679811

**Comment by renukamanavalan** [comment]
@tahmed-dev -- can you please provide your review/approval?

**Comment by tahmed-dev** [comment]
@stepanblyschak, please merge with latest and resolve the conflicts.

**Review by tahmed-dev — APPROVED** [review]

**Comment by stepanblyschak** [comment]
Depends on fix https://github.com/Azure/sonic-buildimage/pull/8044

**Review by tahmed-dev — DISMISSED** [review]

File: Makefile.work [hunk 14]
```diff
@@ -268,6 +268,7 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            HTTPS_PROXY=$(https_proxy) \
                            NO_PROXY=$(no_proxy) \
                            SONIC_INCLUDE_SYSTEM_TELEMETRY=$(INCLUDE_SYSTEM_TELEMETRY) \
+                           INCLUDE_DHCP_RELAY=$(INCLUDE_DHCP_RELAY) \
```
Comment (renukamanavalan) [inline Makefile.work:271]: I see this flag being passed. But where does it gets used ?

File: files/build_templates/init_cfg.json.j2 (hunk 9 above)
Comment (renukamanavalan) [reply files/build_templates/init_cfg.json.j2:49]: How do we ensure that the feature table entry for dhcp_relay has kube as set_owner?

File: rules/docker-dhcp-relay.mk (hunk 3 above)
Comment (renukamanavalan) [reply rules/docker-dhcp-relay.mk:39]: Does this mean, default build of image (say, make target/sonic-broadcom.bin) will not build dhcp_relay and dhcp_relay docker will not be part of final install image ?

File: Makefile.work (hunk 14 above)
Comment (stepanblyschak) [reply Makefile.work:271]: Thanks for catching this, I missed that probably during merges.

**Comment by stepanblyschak** [comment]
@renukamanavalan "How do we ensure that the feature table entry for dhcp_relay has kube as set_owner?"
If INCLUDE_KUBERNETES is y then:
https://github.com/Azure/sonic-buildimage/pull/6531/files#diff-040507c367f31f429105c610d067f0a8df05d33edeef4dc40a7de112a4548ea7R34

and then: https://github.com/Azure/sonic-buildimage/blob/master/files/build_templates/sonic_debian_extension.j2#L689

**Comment by stepanblyschak** [comment]
@renukamanavalan "Does this mean, default build of image (say, make target/sonic-broadcom.bin) will not build dhcp_relay and dhcp_relay docker will not be part of final install image ?"

It is installed by default and enabled, I don't change that.

**Review by ayurkiv-nvda — APPROVED** [review]

File: rules/docker-dhcp-relay.mk [hunk 15]
```diff
@@ -15,17 +15,43 @@ $(DOCKER_DHCP_RELAY)_DBG_IMAGE_PACKAGES = $($(DOCKER_CONFIG_ENGINE_BUSTER)_DBG_I
...
 $(DOCKER_DHCP_RELAY)_LOAD_DOCKERS = $(DOCKER_CONFIG_ENGINE_BUSTER)
 
+$(DOCKER_DHCP_RELAY)_INSTALL_PYTHON_WHEELS = $(SONIC_UTILITIES_PY3)
+$(DOCKER_DHCP_RELAY)_INSTALL_DEBS = $(PYTHON3_SWSSCOMMON)
+
 $(DOCKER_DHCP_RELAY)_VERSION = 1.0.0
 $(DOCKER_DHCP_RELAY)_PACKAGE_NAME = dhcp-relay
+$(DOCKER_DHCP_RELAY)_PACKAGE_DEPENDS = database^1.0.0
 
-SONIC_DOCKER_IMAGES += $(DOCKER_DHCP_RELAY)
```
Comment (renukamanavalan) [inline rules/docker-dhcp-relay.mk:21]: What is the impact of this change ?
Remove from 
SONIC_DOCKER_IMAGES += $(DOCKER_DHCP_RELAY)
SONIC_INSTALL_DOCKER_IMAGES += $(DOCKER_DHCP_RELAY)

**Review by tahmed-dev — APPROVED** [review]

File: rules/docker-dhcp-relay.mk (hunk 15 above)
Comment (renukamanavalan) [reply rules/docker-dhcp-relay.mk:21]: Can you please explain ?
Per prev comments "_It is installed by default and enabled, I don't change that_."
Here it is "_Do not add this docker to a list of regular installed dockers_"

There is a discrepancy. 
Feel free to call me in teams, if that would help

Comment (stepanblyschak) [reply rules/docker-dhcp-relay.mk:21]: @renukamanavalan Sorry, let me clarify this. The key word here is *regular*. It is no longer added to SONIC_INSTALL_DOCKER_IMAGES (as any other docker) but added to SONIC_PACKAGES_LOCAL (new target group handled differently in sonic_debian_extension, basically it will invoke "sonic-package-manager install" for dhcp-relay.)

Comment (renukamanavalan) [reply rules/docker-dhcp-relay.mk:21]: Thanks Stepan!
